testctrl/docs/specs/container_image_names.md

```diff
@@ -0,0 +1,119 @@
+# Container Image Names
+
+The benchmarking framework uses docker containers for each of the test
+components. Since images are built and pushed to a docker registry often, the
+number of images can grow large. Therefore, this spec defines a naming
+convention that can aid automated clean-up scripts.
+
+## Background
+
+Docker images are encapsulated in repositories in a registry. The naming is a
+bit confusing. In terms of the Docker CLI, a tag is `"httpd:v1"` where `"v1"`
+is the version. This differs from the Registry perspective, where `"httpd"` is
+a repository and `"v1"` is a tag.
+
+Docker registries, like Google Container Registry ([gcr.io]) and DockerHub
+([hub.docker.com]), implement a common [Registry API]. However, docker and this
+API do not natively provide batch clean-up actions. Docker also provides no TTL
+on container images.
+
+[gcr.io]: https://gcr.io
+[hub.docker.com]: https://hub.docker.com
+[Registry API]: https://docs.docker.com/registry/spec/api/
+
+One approach for building a TTL on container images would be to use timestamps
+as versions/tags of the images. However, the Registry API cannot query based on
+these. Checking these requires a separate API call for each repository.
+
+Instead of excessive requests, this framework imposes a strict set of naming
+conventions that can aid in automated clean up.
+
+## Specification
+
+The framework allows a repository to contain exactly 1 container image. The
+repository is named as follows:
+
+```
+[ REGISTRY_PREFIX "/" ]
+    "benchmarks" "/"
+          <USER> "/"
+              <TIMESTAMP> "/"
+                  <COMPONENT_KIND> "/"
+                      <LANGUAGE_CODE> [:latest]
+```
+
+1.  The name begins with an optional `REGISTRY_PREFIX`. Some registries, like
+    Google Container Registry (GCR) require a specific prefix. This allows
+    images to be uploaded to these registries.
+
+2.  The word `"benchmarks"` helps organize images. Projects may have many
+    images, and this keyword shows which are related to performance tests. In
+    GCR, these repositories are displayed in a nested form, like a subdirectory.
+
+3.  Each container contains a component's executable that was built after
+    cloning a GitHub repository. The `USER` specifies the username or
+    organization of this repository. For example, the grpc/grpc-java repository
+    would set the user to "grpc". A developer's fork, like codeblooded/grpc,
+    would set the user to "codeblooded".
+
+4.  The `TIMESTAMP` is the time when the set of images were built, represented in
```

Comment:
Btw it looks like that due the the way nested GCR repositories seem to be browsed (need 1 query for each nesting level), the timestamp in the repository name is not going to help you because you'll still need to do queries to discover which images are nested under the timestamp.
Can you doublecheck what the situation is?


https://cloud.google.com/sdk/gcloud/reference/container/images/list

Reply:
I have not personally queried the API yet. In all of the examples I have seen (including [this one](https://www.notsosecure.com/anatomy-of-a-hack-docker-registry/)), they are returned in a non-nested list from `GET /_catalog`.

Reply:
I'd say let's try first so that we don't build the stuff on wrong assumptions.